signals.c

```diff
@@ -0,0 +1,95 @@
+/*
+  Not part of Grbl. KeyMe specific.
+  
+  Signal is one layer of abstraction above adc.h and adc.c.
+
+  ADC values are read in specific time intervals, filtered and
+  stored in the appropriate arrays.
+*/
+
+#include "signals.h"
+#include "adc.h"
+#include "systick.h"
+
+#define N_FILTER 3
+#define SIGNALS_DEFAULT_INTERVAL 1000 
+#define x(i) adc_samples_x[FORCE_VALUE_INDEX][i]
+
+// Unfiltered ADC readings
+int16_t adc_samples_x[VOLTAGE_SENSOR_COUNT][N_FILTER + 1]; 
+
+void signals_init()
+{
+  signals.pause = 0;
+  signals.callback_period = SIGNALS_DEFAULT_INTERVAL;
+  signals.force_offset = 0;
+}
+
+// Update motors ADC readings
+void signals_update_motors()
+{
+  uint8_t idx;
+
+  for (idx = 0; idx < N_AXIS; idx++) {
+    // Assumes the motors are on ADC channels 0-3 and in the same
+    // order in signals.adc_samples. If the pins are changed, a
+    // motors should be mapped to ADC channels.
+    signals.adc_samples[idx] = raw_adc_readings[idx];
+  }
+}
+
+// Filter and update force ADC reading
+void signals_update_force()
+{
+  // Since the raw adc value changes based on an interrupt,
+  // copy the raw value at the beginning of this function to
+  // ensure that the value being used doesn't change during 
+  // during the function call.
+  uint16_t raw_value = raw_adc_readings[FORCE];
+  
+  // Make sure that the offset doesn't make the sample negative.
+  if (signals.force_offset > raw_value) {
+    x(N_FILTER) = 0;
+  } else {
+    x(N_FILTER) = raw_value - signals.force_offset;
+  }
+
+  /*
+    Filter:
+      Moving average Hanning Filter:
+      y[k] = 0.25 * (x[k] + 2x[k-1] + x[k-2])
+  */
+  signals.adc_samples[FORCE_VALUE_INDEX] = (uint16_t)(
+  (x(N_FILTER) + (x(N_FILTER - 1) << 1) + x(N_FILTER - 2)) >> 2);
+
+  // Advance all values in the unfiltered array
+  memmove(&x(0), &x(1) ,sizeof(uint16_t) * N_FILTER-1);
```

Comment:
Sure, but IMO the benefit is not worth the complexity, since changing the filter would result in more complex changes than the current implementation. The way in which it is implemented paves the way for more complex filters, especially if we add another y-signal buffer. Keeping the newest reading at N_FILTER in the x buffer (and in the y buffer in the future) makes it easy to compare the code to the math. N_FILTER is 3, so I think memmove is fine. For our improved filters the  value of N_FILTER will not be more than ~5-10. A value of N_FILTER higher than that, would lower the responsiveness of the ADC reading.